matrix_content_scanner/scanner/scanner.py

```diff
@@ -107,9 +276,38 @@ async def scan_file(
 
         # Raise an error if the result isn't clean.
         if result is False:
-            raise FileDirtyError()
+            raise FileDirtyError(cacheable=cacheable)
 
-        return media
+        return media, cacheable
+
+    def _get_cache_key_for_file(
+        self,
+        media_path: str,
+        metadata: Optional[JsonDict],
+        thumbnail_params: Optional[Dict[str, List[str]]],
+    ) -> str:
+        """Generates the key to use to store the result for the given media in the result
+        cache.
+
+        The key is computed using the media's `server_name/media_id` path, but also the
+        metadata dict (stringified), in case e.g. the decryption key changes, as well as
+        the parameters used to generate the thumbnail if any (stringified), to
+        differentiate thumbnails from full-sized media.
+        The resulting key is a sha256 hash of the concatenation of these two values.
+
+        Args:
+            media_path: The `server_name/media_id` path of the file to scan.
+            metadata: The file's metadata (or None if the file isn't encrypted).
+            thumbnail_params: The parameters to generate thumbnail with. If no parameter
+                is passed, this will be an empty dict. If the media being requested is not
+                a thumbnail, this will be None.
+        """
+        raw_metadata = encode_canonical_json(metadata)
+        raw_params = encode_canonical_json(thumbnail_params)
+        media_path_bytes = media_path.encode("utf8")
+        base_bytes = media_path_bytes + b"\0" + raw_metadata + b"\0" + raw_params
+
+        return hashlib.sha256(base_bytes).hexdigest()
```

Comment:
to avoid building big strings in memory

```suggestion
        hash = hashlib.sha256()
        hash.update(media_path.encode("utf8"))
        hash.update(b"\0")
        hash.update(encode_canonical_json(metadata))
        hash.update(b"\0")
        hash.update(encode_canonical_json(thumbnail_params))
        return hash.hexdigest()
```